File: src/main/java/es/developer/achambi/coreframework/threading/MainExecutor.java
package es.developer.achambi.coreframework.threading;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

public class MainExecutor extends ThreadPoolExecutor {
    private static final long KEEP_ALIVE_TIME = 0;
    private static final int THREAD_NUMBERS = 4;
    private static final Handler MAIN_HANDLER = new Handler( Looper.getMainLooper() );

    private MainExecutor(int corePoolSize, int maximumPoolSize, long keepAliveTime,
                        TimeUnit unit, LinkedBlockingQueue<Runnable> workQueue) {
        super(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue);
    }

    public void executeRequest(final Request request, final ResponseHandler responseHandler ) {
        execute(new Runnable() {
            @SuppressWarnings("ConstantConditions")
            @Override
            public void run() {
                try {
                    postSuccessOnUI( request.perform(), responseHandler );
                } catch (Error e) {
                    postErrorOnUI( e, responseHandler );
                } catch (Exception e) {
                    postErrorOnUI( new Error( e ), responseHandler );
                    e.printStackTrace();
                }
            }
        });
    }

    private void postErrorOnUI( final Error error,
                                final ResponseHandler responseHandler ) {
        MAIN_HANDLER.post(new Runnable() {
            @Override
            public void run() {
                responseHandler.onError( error );
            }
        });
    }

    private void postSuccessOnUI( final Response response,
                                  final ResponseHandler responseHandler ) {
        MAIN_HANDLER.post(new Runnable() {
            @SuppressWarnings("unchecked")
            @Override
            public void run() {
                responseHandler.onSuccess( response );
            }
        });
    }

    public static MainExecutor buildExecutor() {
        return new MainExecutor( THREAD_NUMBERS, THREAD_NUMBERS,
                KEEP_ALIVE_TIME,
                TimeUnit.MILLISECONDS,
                new LinkedBlockingQueue<Runnable>());
    }
}

File: src/main/java/es/developer/achambi/coreframework/threading/MainExecutor.kt
package es.developer.achambi.coreframework.threading

import android.os.Handler
import android.os.Looper
import com.crashlytics.android.Crashlytics
import java.lang.Exception

import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

class MainExecutor private constructor(corePoolSize: Int, maximumPoolSize: Int, keepAliveTime: Long,
                                            unit: TimeUnit, workQueue: LinkedBlockingQueue<Runnable>)
    : ThreadPoolExecutor(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue),
        ExecutorInterface {

    override fun <T> executeRequest(request: Request<T>, responseHandler: ResponseHandler<T>?) {
        execute {
            try {
                postSuccessOnUI(request.perform(), responseHandler)
            } catch (e: CoreError) {
                postErrorOnUI(e, responseHandler)
                Crashlytics.logException(e)
                e.printStackTrace()
            } catch (e:Exception) {
                postErrorOnUI(CoreError(e.toString()), responseHandler)
                Crashlytics.logException(e)
                e.printStackTrace()
            }
        }
    }

    private fun <T> postErrorOnUI(error: CoreError,
                                  responseHandler: ResponseHandler<T>?) {
        MAIN_HANDLER.post { responseHandler?.onError(error) }
    }

    private fun <T> postSuccessOnUI(response: T,
                                responseHandler: ResponseHandler<T>?) {
        MAIN_HANDLER.post { responseHandler?.onSuccess(response) }
    }

    companion object {
        private const val KEEP_ALIVE_TIME: Long = 0
        private const val THREAD_NUMBERS = 4
        private val MAIN_HANDLER = Handler(Looper.getMainLooper())

        fun buildExecutor(): MainExecutor {
            return MainExecutor(THREAD_NUMBERS, THREAD_NUMBERS,
                    KEEP_ALIVE_TIME,
                    TimeUnit.MILLISECONDS,
                    LinkedBlockingQueue())
        }
    }
}
